src/module.c

```diff
@@ -826,23 +828,155 @@ int TSDB_incrby(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
+        }
+
+        ReplyWithSeriesLastDatapoint(ctx, series);
+        RedisModule_CloseKey(key);
+        return REDISMODULE_OK;
+    }
+
+    // Parse timestamp
+    api_timestamp_t timestamp = -1;
+    if (timestamp_location != -1) {
+        long long tmp_timestamp;
+        if ((RedisModule_StringToLongLong(argv[timestamp_location + 1], &tmp_timestamp) !=
+             REDISMODULE_OK) ||
+            tmp_timestamp < 0) {
+            return RTS_ReplyGeneralError(ctx, "TSDB: failed parsing timestamp");
+        }
+        timestamp = (api_timestamp_t)tmp_timestamp;
+    }
+
+    // Parse block time
+    long long block_timeout = -1;
+    if (block_location != -1) { // Blocking waiting for a sample
+        if ((RedisModule_StringToLongLong(argv[block_location + 1], &block_timeout) !=
+             REDISMODULE_OK) ||
+            block_timeout < 0) {
+            return RTS_ReplyGeneralError(ctx, "TSDB: failed parsing block timeout");
+        }
+    }
+
     Series *series;
     RedisModuleKey *key;
-    const int status = GetSeries(ctx, argv[1], &key, &series, REDISMODULE_READ);
-    if (!status) {
-        return REDISMODULE_ERR;
+    const int status = SilentGetSeries(ctx, argv[1], &key, &series, REDISMODULE_READ);
+    // The key exists but not a series
+    if (status == TSDB_ERROR) {
+        return RTS_ReplyGeneralError(ctx, REDISMODULE_ERRORMSG_WRONGTYPE);
     }
 
-    ReplyWithSeriesLastDatapoint(ctx, series);
-    RedisModule_CloseKey(key);
+    // The series exists
+    if (status == TSDB_OK) {
+        // If timestamp wasn't defined the default is the last timestamp
+        if (timestamp == -1) {
+            timestamp = series->lastTimestamp;
+        }
+
+        // In case a retention is set shouldn't return event older than the retention
+        if (series->retentionTime) {
+            timestamp = series->lastTimestamp > series->retentionTime
+                            ? max(timestamp, series->lastTimestamp - series->retentionTime)
+                            : timestamp;
+        }
 
+        // Looking for a sample with timestamp bigger than given timestamp
+        if (timestamp < series->lastTimestamp) {
+            // read next event right after `timestamp`
+            RangeArgs args = { .startTimestamp = timestamp + 1,
+                               .endTimestamp = series->lastTimestamp,
+                               .count = -1,
+                               .aggregationArgs = NULL,
+                               .filterByValueArgs = NULL,
+                               .filterByTSArgs = NULL };
+            SeriesIterator *iterator = SeriesQuery(series, &args, false);
+            if (iterator == NULL) {
+                return RedisModule_ReplyWithArray(ctx, 0);
+            }
+            Sample sample;
+            if (SeriesIteratorGetNext(iterator, &sample) == CR_OK) {
+                ReplyWithSample(ctx, sample.timestamp, sample.value);
+            } // TODO handle error?
```

Comment:
@gkorland ERROR here means there's no sample, you might want to answer with NULL response to make sure there's something being sent to the client.

Reply:
```suggestion
            } else {
                return RedisModule_ReplyWithArray(ctx, 0);
            }
```

Reply:
@gkorland Please fix it